stdlib/public/Concurrency/TaskGroup.swift

```diff
@@ -64,7 +64,7 @@ extension Task {
     resultType: TaskResult.Type,
     returning returnType: BodyResult.Type = BodyResult.self,
     body: (inout Task.Group<TaskResult>) async throws -> BodyResult
-  ) async throws -> BodyResult {
+  ) async rethrows -> BodyResult {
     let task = Builtin.getCurrentAsyncTask()
     let _group = _taskGroupCreate(task: task)
     var group: Task.Group<TaskResult>! = Task.Group(task: task, group: _group)
```

Comment:
A few lines down, this could use defer instead of duplicating the cleanup code in a do/catch

Reply:
Thanks, yeah, I was hitting some crazy "I'm not sure if compiler is doing insane things or if I lost my mind" kinds of situations with this code... so it ended up like this. 

(Situations similar to rdar://74693488 and rdar://74957357 recently made me trust nothing recently...)

I'll give it a shot again and bring back the defer style here, hope it'll be just fine now.

Reply:
For what it's worth we have to rewrite this entire function in C to minimize the exposed ABI for rdar://75063436 🤔 

I'll apply your feedback though, thank you for chiming in! 👍

Reply:
Heh, actually it's a bit annoying because of `'async' call cannot occur in a defer body
      await group._tearDown()` so can't move everything, just some things...

I can do this but not sure it's so much nicer:

```swift

    defer {
      group = nil
      _taskGroupDestroy(task: task, group: _group)
    }

    // Run the withGroup body.
    do {
      let result = try await body(&group)
      await group._tearDown()
      return result
    } catch {
      group.cancelAll()
      await group._tearDown()
      throw error
    }
```

and we'll have to rewrite this into C anyway soon, so maybe just keeping as is; sounds ok @slavapestov ?